osu.Game/Audio/Effects/Filter.cs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using ManagedBass.Fx;
+using osu.Framework.Audio.Mixing;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+
+namespace osu.Game.Audio.Effects
+{
+    public class Filter : Component
+    {
+        public BQFType FilterType = BQFType.LowPass;
+        public float SweepCutoffStart = 2000;
+        public float SweepCutoffEnd = 150;
+        public float SweepDuration = 100;
+        public Easing SweepEasing = Easing.None;
+
+        public bool IsActive { get; private set; }
+
+        private readonly Bindable<float> filterFreq = new Bindable<float>();
+        private readonly AudioMixer mixer;
+        private BQFParameters filter;
+
+        /// <summary>
+        /// A BiQuad filter that performs a filter-sweep when toggled on or off.
+        /// </summary>
+        /// <param name="mixer">The mixer this effect should be attached to.</param>
+        public Filter(AudioMixer mixer)
+        {
+            this.mixer = mixer;
+        }
+
+        public void Enable()
+        {
+            attachFilter();
+            this.TransformBindableTo(filterFreq, SweepCutoffEnd, SweepDuration, SweepEasing);
+        }
+
+        public void Disable()
+        {
+            this.TransformBindableTo(filterFreq, SweepCutoffStart, SweepDuration, SweepEasing).OnComplete(_ => detatchFilter());
+        }
+
+        private void attachFilter()
+        {
+            if (IsActive) return;
+
+            filter = new BQFParameters
+            {
+                lFilter = FilterType,
+                fCenter = filterFreq.Value = SweepCutoffStart
+            };
```

Comment:
Following up on the discord discussion about the cutoff behaving strangely near nyquist, [Ian from un4seen has advised](https://www.un4seen.com/forum/?topic=19542.msg136677#msg136677) to use the following params instead:

```suggestion
            filter = new BQFParameters
            {
                lFilter = FilterType,
                fCenter = filterFreq.Value = SweepCutoffStart,
                fBandwidth = 0,
                fQ = 0.7f
            };
```

I've tested with the [`filter-wip` branch](https://github.com/nekodex/osu/tree/filter-wip) you had and it looks to work.

(Although, with the way you have things set up on that branch, do note that it will be initially silent because of how the `Filter` logic looks like after removal of `Enable()`/`Disable()` - `Filter` starts with a cutoff freq of nyquist - 1Hz, which applied to the high pass filter just mutes everything; that was masked by the bass_fx-side overflow. I just removed the high pass/band pass filters to test but you'd probably need to bring the methods back or use a saner default depending on the type of filter.)